src/avl_tree.jl

```diff
@@ -0,0 +1,259 @@
+            else
+                node.rightChild = right_rotate(node.rightChild)
+                return left_rotate(node)
+            end
+        end
+
+        return node
+    end
+
+    haskey(tree, d) && return tree
+
+    tree.root = insert_node(tree.root, d)
+    tree.count += 1
+    return tree
+end
+
+"""
+    push!(tree, key)
+
+Inserts `key` in the `tree` if it is not present.
+"""
+function Base.push!(tree::AVLTree{K}, key0) where K
+    key = convert(K, key0)
+    insert!(tree, key)
+end
+
+function Base.delete!(tree::AVLTree{K}, d::K) where K
+
+    function delete_node(node::Union{AVLTreeNode, Nothing}, key)
+        if node == nothing || node.data == nothing
+            return nothing
+        elseif key < node.data
+            node.leftChild = delete_node(node.leftChild, key)
+        elseif key > node.data
+            node.rightChild = delete_node(node.rightChild, key)
+        else
+            if node.leftChild == nothing
+                temp = node.rightChild
+                node = nothing
+                return temp
+            elseif node.rightChild == nothing
+                temp = node.leftChild
+                node = nothing
+                return temp
+            else 
+                temp = minimum_node(node.rightChild)
+                node.data = temp.data
+                node.rightChild = delete_node(node.rightChild, temp.data)
+            end 
+        end
+
+        node.height = fix_height(node)
+        balance = get_balance(node)
+
+        if balance > 1
+            if get_balance(node.leftChild) >= 0
+                return right_rotate(node)
+            else
+                node.leftChild = left_rotate(node.leftChild)
+                return right_rotate(node)
+            end
+        end
+
+        if balance < -1
+            if get_balance(node.rightChild) <= 0
+                return left_rotate(node)
+            else
+                node.rightChild = right_rotate(node.rightChild)
+                return left_rotate(node)
+            end
+        end 
+
+        return node
+    end
+
+    # if the key is not in the tree, do nothing and return the tree
+    !haskey(tree, d) && return tree
+    
+    # if the key is present, delete it from the tree
+    tree.root = delete_node(tree.root, d)
+    tree.count -= 1
+    return tree
+end
+
+"""
+    getindex(tree, ind)
+
+Gets the key present at index `ind` of the tree. Indexing is done in increasing order of key.
+"""
+getindex(tree, ind)
+
+function Base.getindex(tree::AVLTree{K}, ind) where K 
+    @boundscheck (1 <= ind <= tree.count) || throw(ArgumentError("$ind should be in between 1 and $(tree.count)"))
+    function traverse_tree_inorder(node::Union{AVLTreeNode, Nothing})
+        if (node != nothing)
+            left = traverse_tree_inorder(node.leftChild)
+            right = traverse_tree_inorder(node.rightChild)
+            append!(push!(left, node.data), right)
+        else
+            return K[]
```

Comment:
Can this ever happen?
I don't think it can since we did a check

Reply:
Yes. It can. `node` might not be `nothing`, but it's `leftChild` and `rightChild` can be.

Reply:
this will cause a out of bounds error a few lines later right?
Should we just throw on directly here?

Reply:
Ya, we could. Even I was thinking about that. 
Once `traverse_tree_inorder` is called, we can be sure that, there is a valid index. So, it will come from the `leftChild`, or the `rightChild`. 
Let `idx` be the index we are interested in. 
`L` = size of the `left` array
`R` = size of the `right` array.
```julia
if idx <= L
    return left[idx]
else
    return right[idx-L]
```